### PR TITLE
EVG-20073 Create versions for GitHub merge queue

### PR DESCRIPTION
**Pull request by bsamek**
EVG-20073

### Description

This adds support to the patch-intent-processor amboy job for GitHub merge queue
items.

### Testing

This tests the new requester type by running similar integration tests for the
patch-intent-processor job as currently exist for other requester types.

### Documentation

Light edits to the ADR. There are not user-facing changes yet, so I have not
added public docs.

### CODE REVIEW

File: model/patch/github_merge.go [hunk 1]
```diff
@@ -47,28 +48,42 @@ type githubMergeIntent struct {
...
+	}
+	if mg.GetRepo().GetName() == "" {
+		return nil, errors.New("merge group repo name cannot be empty")
+	}
+	if headRef := mg.GetMergeGroup().GetHeadRef(); headRef == "" {
 		return nil, errors.New("merge group head ref cannot be empty")
 	}
-	if mg.GetHeadSHA() == "" {
+	if mg.GetMergeGroup().GetHeadSHA() == "" {
 		return nil, errors.New("head ref cannot be empty")
```
Comment (malikchaya2) [inline model/patch/github_merge.go:77]: Should this say head `sha`?

**Review by malikchaya2 — CHANGES_REQUESTED** [review]

File: model/patch/github_merge.go [hunk 2]
```diff
@@ -140,13 +155,32 @@ func (g *githubMergeIntent) GetCalledBy() string {
...
+
+	// handle cases where base branch has a slash in it
+	baseBranchSlice := []string{}
+	for i := 3; i < len(split)-1; i++ {
+		baseBranchSlice = append(baseBranchSlice, split[i])
+	}
+	baseBranch := strings.Join(baseBranchSlice, "/")
+
+	// produce a branch name like gh-readonly-queue/main/pr-515-9cd8a2532bcddf58369aa82eb66ba88e2323c056
+	headBranch := strings.Join([]string{split[2], baseBranch, split[len(split)-1]}, "/")
```
[the remark this answers is not in the document]
Comment (bsamek) [reply model/patch/github_merge.go]: done

File: model/patch/patch.go [hunk 3]
```diff
@@ -772,8 +772,15 @@ func (p *Patch) IsPRMergePatch() bool {
 	return p.GithubPatchData.MergeCommitSHA != ""
 }
 
+// IsCommitQueuePatch returns true if the the patch is part of any commit queue:
+// either Evergreen's commit queue or GitHub's merge queue.
 func (p *Patch) IsCommitQueuePatch() bool {
-	return p.Alias == evergreen.CommitQueueAlias || p.IsPRMergePatch()
+	return p.Alias == evergreen.CommitQueueAlias || p.IsPRMergePatch() || p.GithubMergeData.HeadSHA != ""
```
[the remark this answers is not in the document]
Comment (bsamek) [reply model/patch/patch.go]: done

File: units/patch_intent.go [hunk 4]
```diff
@@ -998,6 +1050,26 @@ func findEvergreenUserForPR(githubUID int) (*user.DBUser, error) {
...
 }
 
+func findEvergreenUserForGithubMergeGroup(githubUID int) (*user.DBUser, error) {
+	u, err := user.FindOne(user.ById(evergreen.GithubMergeUser))
+	if err != nil {
+		return u, errors.Wrap(err, "finding GitHub merge queue user")
+	}
+	// and if that user doesn't exist, make it
+	if u == nil {
+		u = &user.DBUser{
```
[the remark this answers is not in the document]
Comment (bsamek) [reply units/patch_intent.go:1060]: I think it's better for this to be in code than just in the database. If it's only in the database, then we could try to start another Evergreen (for example, a staging environment), and it won't start until someone figures out what the error message means. I'm also following the convention in the function just above this one, for the GitHub PR user.

[the remark this answers is not in the document]
Comment (malikchaya2) [reply units/patch_intent.go:1060]: good point, thanks!

**Review by malikchaya2 — APPROVED** [review]

**Pull request merged**